### DISCUSSION TITLE
About SunSpec Modbus protocol

### DISCUSSION THREAD

**Discussion by mi-hol** [gavinying/modpoll#83] [Q&A]
> Maybe a test case using modsim.topmaker.net device could be added for a 16 characters long string?

Ideal for tests would be to extract:

```
            "VendorUrl": "https://github.com/riptideio/pymodbus/",
            "ProductName": "Pymodbus Server",
```

_Originally posted by @mi-hol in https://github.com/gavinying/modpoll/issues/77#issuecomment-2549020964_

**Comment by mi-hol** [comment]
and field "MajorMinorRevision": "3.8.0"

**Comment by gavinying** [comment]
In fact, the register's string length shall be according to device specification, it is not configured or decided by modpoll tool.

For your request, we need to modify modsim code to simulate a 16-character long string value, see https://github.com/gavinying/modsim/commit/392113d7586cc0d89fb40ef40ff142b116b30a79#diff-1214100fc30f036ad98e2a546fcb20b6fb7a7cf375bff7fcbb4a099590f6cbfdR97

Now, you could use below config to retrieve 16-byte long string value from modsim device,
```
device,modsim,1,,
poll,holding_register,40036,8,BE_BE
ref,holding_reg19,40036,string16,rw
```
I will later update the modpoll example config files to include above usages as well. thanks. 

Btw, the modsim is only a very simple example of Modbus simulator (all logics are in one python file), you could fork it and easily make changes to adapt to your further tests.

**Comment by mi-hol** [comment]
> I will later update the modpoll example config files to include above usages as well.

Thanks, I see my misunderstanding now.
"Poll" extracts 8 words, which is containing  the  16-character long string value.
did I get this right?

**Comment by gavinying** [comment]
Sorry for the late response, you are absolutely right, 16-character long string is 8 words, so here we shall poll 8 registers in the config file. 
Hope you have already get the problem solved.

**Reply by mi-hol** [reply]
> Sorry for the late response, you are absolutely right, 16-character long string is 8 words, so here we shall poll 8 registers in the config file. Hope you have already get the problem solved.

From my understanding polling 16 words in one poll definition should allow for referencing 2 strings each 16 bytes long but  
Warning: ` W | modpoll.modbus_task | Reference Model failed to pass sanity check, ignoring it.` is output from using below definition.
Likely  I'm doing again something wrong due to lack of understanding basics :(

ModBus implementation used is below in https://github.com/gavinying/modpoll/discussions/83#discussioncomment-11760727
modpoll Definition:

```
device,FroniusGen24Sim,1,,

poll,holding_register,40005,16,BE_BE
ref,Manufacturer,40005,string16,r
ref,Model,40021,string16,r

```

Proposed solution:

Provide more details, so a newbie can understand "how to fix" the warning "failed to pass sanity check"

**Reply by mi-hol** [reply]
Note: this warning was from version 1.4.0 but I assume 1.4.1 was not changed in that area of code

**Comment by mi-hol** [comment]
> Btw, the modsim is only a very simple example of Modbus simulator (all logics are in one python file), you could fork it and easily make changes to adapt to your further tests

To clarify my intend, my request is targeting to improve usability of modpoll for all possible use cases as it is such a great tool! 
In this specific case no example for extracting strings exists in current documentation.

So far I found modpoll to work perfectly in environments where device specifications are fully correct, but errors in specifications and implementations are are a fact of life.
I'll add more issues around handling incorrect output of modpoll **due to errors in creator supplied device specifications & implementations** shortly.

**Comment by gavinying** [comment]
I have to agree that the current document is far from covering all features or potential usecases of `modpoll` tool. the tool itself is quite flexible and can be used in many senarios, I'll continuously improve our documents. Thanks for your suggestions. 

Meanwhile, I would also like to say, I always **welcome contributions from users**, because people love to see and learn from actual examples, you might have just come through a hard time (due to non-standard vendor implementation or even incorrect vendor specs) to figure out a modpoll config for a specific device, now the other people happens to use the same vendor might be benefit from your work if you could share it. It could be a config file or any writeups of your experience or tips. **PULL REQUESTS ARE WELCOME**.

In fact, this is the gist of creating this `modpoll` tool, which made it different from the other traditional non-sharable `modpoll` tools. 

Happy coding and sharing :)

**Comment by mi-hol** [comment]
> I always **welcome contributions from users**

I plan to contribute like I did in many other OSS projects already.
Unfortunately I've not mastered ModBus yet but modpoll is helping a lot.

I try to read data from devices supporting [SunSpec](https://sunspec.org/sunspec-technology-overview/) but have to rely currently on an ESP32-based simulator. This emulator is written in Tasmota script and suffers from a number of issues I hope to fix.
To complicate matters further, the emulator's source code is currently only available on a [file sharing website](https://pastebin.com/11zX4kG5) with [description on a different website](https://ottelo.jimdofree.com/stromz%C3%A4hler-auslesen-tasmota/) :(
I'm working with the original author to publish on GH but with limited success yet.

The emulator's ModBus related code is below

current power in W = 40097 
total Wh export = 40129 
total Wh Import = 40137 

```
switch reg
  case 768
        data={0}
 
    case 1706
        data={0}
 
    case 40000
        data={21365 28243}
 
    case 40002
        data={1 65}
 
    case 40004
        data={70 114 111 110 105 117 115 0 0 0 0 0 0 0 0 0}
        data={83 109 97 114 116 32 77 101 116 101 114 32 54 51 65 0}
        data={0 0 0 0 0 0 0 0}
        data={0 0 0 0 0 0 0 0}
        data={48 48 48 48 48 48 48 49 0 0 0 0 0 0 0 0}
        data={240 213}
 
    case 40069
        data={213 124}
 
    case 40071
        data={0 0 0 0 0}
        data={0 0 0 0 0}
        data={0 0 0 vsml1 0}
        data={0 0 0 0 0}
        data={0 0 0 0 0}
        data={0 0 0 0 0}
        dform=3
        nval=nval/2
 
    case 40129
        data={vsml3 0 0 0 vsml2}
        data={0 0 0 0 0}
        data={0 0 0 0 0}
        data={0}
        dform=3
        nval=nval/2
 
    case 40193
        data={0 0}
 
    case 40195
        data={65535 0}
 
    case 50000
        data={0 0}
 
```

**Comment by gavinying** [comment]
Do you have the SunSpec Modbus protocol? why not follow the document straight away? also interested to know if your goal is to read data from actual SunSpec device, or to create a SunSpec emulator?

**Reply by mi-hol** [reply]
> Do you have the SunSpec Modbus protocol? 

no, currently only a vendor specification that claims to conform to SunSpec Modbus protocol

> why not follow the document straight away? 

Confusing is the 0 vs 1 based register address numbering used in documents.
Therefore a common error is reading data from an "off by 1" address that seems to create wrong outputs from modpoll

>also interested to know if your goal is to read data from actual SunSpec device, or to create a SunSpec emulator?

both, with sequence: 

1.  SunSpec emulator first, due to testability
2.  verify with multiple SunSpec devices
3.  fix and enhance SunSpec emulator for vendor specifics

**Reply by gavinying** [reply]
Suggest to follow the official SunSpec documents if you want a proper "SunSpec emulator". Don't just rely on vendor's docs and code, unless a vendor specific emulator is your goal. 
In this case, `modpoll` tool can be a test client to help your "emulator" development, just create the modpoll config according to either SunSpec or vendor-specific protocol.
